Initial Classes

mecha opened the pull request.

mecha committed:
Updated dependencies

mecha committed:
Added `AbstractSettingsElement`

mecha committed:
Added `dhii/human-readable-helper-base` dependency

mecha committed:
`AbstractSettingsElement` uses `LabelAwareTrait`

mecha committed:
Added render context normalization

mecha committed:
Added `CouldNotRenderException` factory method

mecha committed:
Added `AbstractField`

mecha committed:
Added render failure test for `AbstractField`

mecha committed:
Removed unused closure variable use

mecha committed:
Fixed code standards

mecha committed:
Added `CompositeValidatorTrait`

mecha committed:
Updated dependencies

mecha committed:
Added `AbstractSection`

mecha committed:
Removed field render mock functionality

mecha committed:
Added assertion failure messages

mecha committed:
Added fields render expectations in abstract section test

mecha committed:
Removed useless exception expectation

mecha committed:
Default timezone for tests to avoid warnings

mecha committed:
Fixed expectations for `AbstractSection` test

mecha committed:
`AbstractSettingsElement` extends `AbstractValidator`

mecha committed:
Code cleanup in `AbstractSection` test

XedinUnknown commented on src/CompositeValidatorTrait.php:
```diff
@@ -0,0 +1,68 @@
+        $errors = [];
+
+        foreach ($this->_getChildrenValidators() as $_key => $_validator) {
+            $errors[$_key] = $this->_getValidationErrorsForChild($subject, $_validator);
```
Of course, it's very useful to be able to get the errors by validator. However, consumers of the results expect a 1-dimentional array, i.e. a list, of errors, not a list of lists. Therefore, this introduces a bug.

How about adding a 2nd parameter e.g. `$isGroupByChild`, `false` by default, which, if true, would do this; otherwise, would flatten everything?

XedinUnknown commented on src/CompositeValidatorTrait.php:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @return ValidatorInterface[]|Traversable The validator instances, mapped via unique keys.
+     */
+    abstract protected function _getChildrenValidators();
```
How about `_getChildValidators()` instead?

XedinUnknown commented on src/AbstractField.php:
```diff
@@ -0,0 +1,112 @@
+            );
+        }
+
+        return sprintf(
```
Didn't really like this. Maybe, we can make it use a template?

mecha replied:
Sure. What's the best approach we can take?

XedinUnknown commented on src/AbstractField.php:
```diff
@@ -0,0 +1,112 @@
+    protected function _renderLabelCell($context)
+    {
+        return sprintf(
+            '<th scope="row"><label for="%1$s">%2$s</label></th>',
```
Also better template?

XedinUnknown commented on src/AbstractField.php:
```diff
@@ -0,0 +1,112 @@
+    protected function _renderContentCell($context)
+    {
+        return sprintf(
+            '<td>%1$s <p class="description">%2$s</p></td>',
```
Template.

XedinUnknown commented on src/AbstractSection.php:
```diff
@@ -0,0 +1,206 @@
+        $heading = $this->_renderHeading($context);
+        $content = $this->_renderContent($context);
+
+        return sprintf('%1$s %2$s', $heading, $content);
```
Template?

XedinUnknown commented on src/AbstractSection.php:
```diff
@@ -0,0 +1,206 @@
+     */
+    protected function _renderHeading(ContainerInterface $context = null)
+    {
+        return sprintf('<h2>%s</h2>', $this->_getLabel());
```
Template?

XedinUnknown commented on src/AbstractSection.php:
```diff
@@ -0,0 +1,206 @@
+    {
+        $rFields = $this->_renderAllFields($context);
+        $table   = sprintf(
+            '<table id="%1$s" class="form-table"><tbody>%2$s</tbody></table>',
```
Template?

mecha committed:
Fixed test for `AbstractField`

mecha committed:
Renamed `_getChildrenValidators` to `_getChildValidators`

mecha committed:
Composite validator does not group errors by default

mecha committed:
Added context factory in `AbstractSectionTest`

mecha committed:
Fixed func test for `AbstractSettingsElement`

mecha committed:
Added `AbstractScreen` and functional test

mecha committed:
Updated `rebelcode/wp-settings-interface`

mecha committed:
Added `TypeAwareTrait`

mecha committed:
Added `BlockAwareTrait`

mecha committed:
`AbstractSettingsElement` has basic rendering and ...

... is no longer a validator.

mecha committed:
Refacted `BlockAwareTrait` into ...

... `BlockAwareSettingsElementTrait` to be specific to settings elements that render via a block.

mecha committed:
Removed unused trait from `AbstractSettingsElement`

mecha committed:
Renamed `AbstractSettingsElement` to `AbstractElement`

mecha committed:
Removed current code for revision of concepts

mecha committed:
Updated dependencies

mecha committed:
Fixed mock type-hint in `TypeAwareTrait` test

mecha committed:
Updated dependencies using PHP 5.4 environment

mecha committed:
Added `Setting` implementation and test

mecha committed:
Added `SettingTypeAwareTrait` and test

mecha committed:
Added `SettingAwareTrait` and test

mecha committed:
Setting type can be a stringable

mecha committed:
Fixed invalid dependency `dhii/data-key-value-abstract`

mecha committed:
Updated dependencies

mecha committed:
Added `AbstractSettingBlock` and test

mecha committed:
Added `AbstractDelegateSettingBlock` and test

mecha committed:
Added `DelegateSettingBlock` and test

mecha committed:
Added `AbstractBaseSettingBlock` and test

mecha committed:
Added `AbstractContainerDelegateSettingBlock` and test

mecha committed:
Updated dependencies

mecha committed:
Updated dependencies

mecha committed:
Added `AbstractBaseSettingBlock#__toString()`

mecha committed:
Removed `AbstractContainerDelegateSettingBlock` ...

... Purpose of the class is not in the scope of this package.

mecha committed:
Removed `AbstractBaseSettingBlock`

mecha committed:
Made `AbstractDelegateSettingBlock` more granular

The delegation is now specific to the rendering of the setting's element.
This allows settings blocks to preserve consistency for rendering the wrapping element, and only delegate the rendering of the most dynamic part.

mecha committed:
Removed `DelegateSettingBlock`. No longer relevant

mecha committed:
Removed old `DelegateSettingBlockTest`

mecha committed:
Refacted methods in `AbstractDelegateSettingBlock`

mecha committed:
Added test for `AbstractDelegateSettingBlock`

mecha committed:
Updated dependencies
